### PR TITLE
POC SSL Certificate

### PR DESCRIPTION
**Pull request by chadlagore**
:hourglass: **Status**: WIP

:tickets: **Ticket(s)**: #59

---

## :flashlight: Testing Instructions

```bash
./gen-cert.sh  # lots of input - ideally pipe this script over ssh and fill in with defaults
sudo go run main.go  # port 443 requires sudo, bruno will be pissed
```

Elsewhere...

```bash
$ curl -sL https://localhost:443
```

Produces server-side error:

```bash
2018/01/18 21:00:43 http: TLS handshake error from [::1]:53606: remote error: tls: unknown certificate authority
```

@jordanschalm and I discuss:

> ... because we control both the client and the server, and we have to ssh pipe, we can just provide any identity that both sides can agree on. We don’t actually care about identities here, all we want is the encryption, so we can kind of fudge the identity part.

Discuss?

### CODE REVIEW

**Comment by jordanschalm** [comment]
@bfbachmann I don't think that's a concern. Since the bootstrapping connection occurs over a secure channel we can inform the client of what cert the server has generated, so the client knows what it should trust. 

The cert *does* come from a trusted authority (our code, running on the server), but the library doesn't know that without some extra work on our part. That's the hard part.

File: poc/tls/gen-cert.sh [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+openssl genrsa -out server.key 2048
```
Comment (jordanschalm) [inline poc/tls/gen-cert.sh]: Can we remove the dependency on openssl?

File: poc/tls/main.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package main
+
+import (
+	"log"
+	"net/http"
+)
+
+func HelloServer(w http.ResponseWriter, req *http.Request) {
```
Comment (jordanschalm) [inline poc/tls/main.go]: Hahaha damn this is simple 😆

**Comment by chadlagore** [comment]
May be of use https://godoc.org/github.com/spacemonkeygo/openssl

File: poc/tls/gen-cert.sh (hunk 1 above)
Comment (chadlagore) [reply poc/tls/gen-cert.sh]: May be of use https://godoc.org/github.com/spacemonkeygo/openssl

**Comment by chadlagore** [comment]
No @bobheadxi, this will be handled over SSH prior to booting the daemon container 😄

**Pull request closed without merging**